### PR TITLE
Updated timeseries_dataset_from_array

### PR DESCRIPTION
**Pull request by markub3327**
Hello,

 @divyashreepathihalli @qlzh727 @gowthamkpr  solution of #16315

   I added a feature for seq-to-seq, too. If targets are defined for each step in window, this is a useful feature.

Thanks.

### CODE REVIEW

**Comment by markub3327** [comment]
I created a unit test for the `element_shape` as is mentioned in the issue.

File: keras/utils/timeseries_dataset.py [hunk 1]
```diff
@@ -73,6 +71,10 @@ def timeseries_dataset_from_array(
     end_index: Optional int; data points later (exclusive) than `end_index`
       will not be used in the output sequences.
       This is useful to reserve part of the data for test or validation.
+    seq_to_seq: bool; whether the target
```
Comment (qlzh727) [inline keras/utils/timeseries_dataset.py:74]: It is not very clear to me what's the usage for this flag.

File: keras/utils/timeseries_dataset.py [hunk 2]
```diff
@@ -128,6 +130,23 @@ def timeseries_dataset_from_array(
   X = np.arange(100)
   Y = X*2
 
+  sample_length = 20
+  dataset = tf.keras.preprocessing.timeseries_dataset_from_array(
+    X, Y, sequence_length=sample_length, sequence_stride=sample_length, seq_to_seq=True)
```
Comment (qlzh727) [inline keras/utils/timeseries_dataset.py:135]: when user have seq_to_seq == False in this case, what will be the output here?

File: keras/utils/timeseries_dataset.py (hunk 1 above)
Comment (qlzh727) [inline keras/utils/timeseries_dataset.py:74]: what's the dfeault value for this flag?

File: keras/utils/timeseries_dataset.py [hunk 3]
```diff
@@ -189,41 +208,21 @@ def timeseries_dataset_from_array(
...
-          lambda i, positions: tf.range(  # pylint: disable=g-long-lambda
-              positions[i],
-              positions[i] + sequence_length * sampling_rate,
-              sampling_rate),
-          num_parallel_calls=tf.data.AUTOTUNE)
-
-  dataset = sequences_from_indices(data, indices, start_index, end_index)
+
+  dataset = tf.data.Dataset.from_tensor_slices(data[start_index : end_index])
+  dataset = dataset.window(sequence_length, shift=sequence_stride, stride=sampling_rate, drop_remainder=True)
```
Comment (qlzh727) [inline keras/utils/timeseries_dataset.py:213]: why are we forcing the drop_remainder to be True here?

File: keras/utils/timeseries_dataset.py [hunk 4]
```diff
@@ -189,41 +208,21 @@ def timeseries_dataset_from_array(
...
+  dataset = dataset.flat_map(lambda x: x.batch(sequence_length, drop_remainder=True))
```
Comment (qlzh727) [inline keras/utils/timeseries_dataset.py:214]: I am bit confused about why we need this here.

File: keras/utils/timeseries_dataset.py [hunk 5]
```diff
@@ -189,41 +208,21 @@ def timeseries_dataset_from_array(
...
-        (tf.data.Dataset.range(len(start_positions)), positions_ds)).map(
-            lambda i, positions: positions[i],
-            num_parallel_calls=tf.data.AUTOTUNE)
-    target_ds = sequences_from_indices(
-        targets, indices, start_index, end_index)
+    if seq_to_seq:
+      target_ds = tf.data.Dataset.from_tensor_slices(targets[start_index : end_index])
+      target_ds = target_ds.window(sequence_length, shift=sequence_stride, stride=sampling_rate, drop_remainder=True)
+      target_ds = target_ds.flat_map(lambda x: x.batch(sequence_length, drop_remainder=True))
+    else:
```
Comment (qlzh727) [inline keras/utils/timeseries_dataset.py:221]: It is unclear to me what the else case will achieve here.

**Review by qlzh727 — CHANGES_REQUESTED** [review]

File: keras/utils/timeseries_dataset.py (hunk 1 above)
Comment (markub3327) [reply keras/utils/timeseries_dataset.py:74]: It's like many-to-many scenario but without zip it manually.

File: keras/utils/timeseries_dataset.py (hunk 2 above)
Comment (markub3327) [reply keras/utils/timeseries_dataset.py:135]: This use case is logically only if seq_to_seq = True

File: keras/utils/timeseries_dataset.py (hunk 1 above)
Comment (markub3327) [reply keras/utils/timeseries_dataset.py:74]: = False

File: keras/utils/timeseries_dataset.py (hunk 3 above)
Comment (markub3327) [reply keras/utils/timeseries_dataset.py:213]: Because I need the same window size in all batches !!!

File: keras/utils/timeseries_dataset.py (hunk 4 above)
Comment (markub3327) [reply keras/utils/timeseries_dataset.py:214]: I think it's good. I do all unit tests. See the tutorial to TF Dataset.

`dataset = dataset.flat_map(lambda x: x.batch(sequence_length, drop_remainder=True))` = this part with `drop_remainder=True` makes the issue #16315 solved because XLA makes shape static and is defined in all axis !

**Comment by markub3327** [comment]
I delete the `seq_to_seq` flag.

File: keras/utils/timeseries_dataset.py (hunk 3 above)
Comment (markub3327) [reply keras/utils/timeseries_dataset.py:194]: @qlzh727 You have true only if I active `drop_remainder=True` at line 208 in `dataset = dataset.batch(batch_size) `. But this is another thing.

Comment (qlzh727) [reply keras/utils/timeseries_dataset.py:194]: I don't think the model need to have the same sequence length all the time. Its very common to have uneven sequence length dataset, and this is why the drop_remainder is a user configurable flag for tf.data API. See https://www.tensorflow.org/api_docs/python/tf/data/Dataset?version=nightly#window for the uneven example.

Comment (markub3327) [reply keras/utils/timeseries_dataset.py:194]: @qlzh727  
Ok. I can make a `drop_remainder` flag to `timeseries_dataset_from_array` or close this PR whole.

Thanks.

**Comment by markub3327** [comment]
@qlzh727 

I added the `drop_remainder` flag.

Thanks.

File: keras/utils/timeseries_dataset.py (hunk 3 above)
Comment (markub3327) [reply keras/utils/timeseries_dataset.py]: @qlzh727 the original code has same sequence length in whole dataset.

Look here: https://github.com/keras-team/keras/blob/master/keras/utils/timeseries_dataset.py#L215

**Comment by markub3327** [comment]
@qlzh727

I changed the `seq_to_seq` terminology to `many_to_many` as is described in the example 3 in docstring.

I point to Option 1 and Option 2 in the example as the thing with the same result, but cleaner code.

Thanks.

**Comment by markub3327** [comment]
@fchollet 

Please, Can you review it?

**Comment by gbaned** [comment]
@markub3327 Can you please resolve conflicts? Thank you!

**Comment by markub3327** [comment]
@gbaned 

 Please concretize conflicts. It's not so clear to me. Thanks.

**Comment by gbaned** [comment]
Hi @markub3327  Can you please resolve conflicts here https://github.com/keras-team/keras/pull/16533/conflicts ? Thank you!

**Comment by markub3327** [comment]
Hi @gbaned all is done. You can merge.

File: keras/utils/timeseries_dataset.py [hunk 6]
```diff
@@ -69,6 +67,12 @@ def timeseries_dataset_from_array(
       shuffle: Whether to shuffle output samples,
         or instead draw them in chronological order.
       seed: Optional int; random seed for shuffling.
+      drop_remainder: Optional bool; representing whether the last windows
```
Comment (fchollet) [inline keras/utils/timeseries_dataset.py:70]: The argument makes sense.

File: keras/utils/timeseries_dataset.py [hunk 7]
```diff
@@ -69,6 +67,12 @@ def timeseries_dataset_from_array(
...
+        should be dropped if their size is smaller than size.
+      many_to_many: Optional bool; whether the target
```
Comment (fchollet) [inline keras/utils/timeseries_dataset.py:72]: This seems like a fairly confusing API, please remove.

**Review by fchollet — COMMENTED** [review]
Thanks for the PR. I suggest only keeping the `drop_remainder` argument. For the many-to-many use case, I recommend rolling out your own functionality separately (outside the core API). Thank you!

**Review by fchollet — COMMENTED** [review]
Thanks for the PR. I suggest only keeping the `drop_remainder` argument. For the many-to-many use case, I recommend rolling out your own functionality separately (outside the core API). Thank you!

File: keras/utils/timeseries_dataset.py (hunk 6 above)
Comment (markub3327) [reply keras/utils/timeseries_dataset.py:70]: Ok!

File: keras/utils/timeseries_dataset.py (hunk 7 above)
Comment (markub3327) [reply keras/utils/timeseries_dataset.py:72]: Ok!

**Comment by markub3327** [comment]
Now many_to_many functionality is provided by the old procedure .... original Example 3.

File: keras/utils/timeseries_dataset.py [hunk 8]
```diff
@@ -212,47 +214,29 @@ def timeseries_dataset_from_array(
...
-        (tf.data.Dataset.range(len(start_positions)), positions_ds)
-    ).map(
-        lambda i, positions: tf.range(
-            positions[i],
-            positions[i] + sequence_length * sampling_rate,
-            sampling_rate,
-        ),
-        num_parallel_calls=tf.data.AUTOTUNE,
+    dataset = tf.data.Dataset.from_tensor_slices(data[start_index:end_index])
+    dataset = dataset.window(
```
Comment (fchollet) [inline keras/utils/timeseries_dataset.py:219]: This is a big departure from the previous implementation. Could you provide a quick performance benchmark of each version on a large dataset of long timeseries?

Comment (markub3327) [reply keras/utils/timeseries_dataset.py:219]: I don't have enough computation power and large dataset. Please can you provide it by yourself? Thanks a lot for your time.

Comment (haifeng-jin) [reply keras/utils/timeseries_dataset.py:219]: Anyone from the community is welcome to continue this work to do the benchmarks. We do not have immediate bandwidth to do the benchmarking.

**Comment by gbaned** [comment]
Hi @fchollet / @qlzh727 Any update on this PR? Please. Thank you!

**Comment by markub3327** [comment]
@fchollet @qlzh727 

Here is the benchmark: https://gist.github.com/markub3327/7534c726208811e2312445455c7884c0

**Comment by mattdangerw** [comment]
Took a look at the benchmark. Is looks like `timeseries_dataset_from_array2`, which I guessing is the new version, is actually performing noticeably worse. Is that correct? We will probably not be able to land this with a performance regression.

**Comment by gbaned** [comment]
Hi @markub3327 Any update on this PR? Please. Thank you!

**Comment by markub3327** [comment]
No..... my solution is using clearer code and more methods from TF.Dataset API, but it was slow.

**Comment by sachinprasadhs** [comment]
Hello, Thank you for submitting a pull request.

We're currently in the process of migrating the new `Keras 3 code base` from `keras-team/keras-core` to `keras-team/keras`. 
Consequently, merging this PR is not possible at the moment. After the migration is successfully completed, feel free to reopen this PR at `keras-team/keras` if you believe it remains relevant to the `Keras 3 code base`. If instead this PR fixes a bug or security issue in legacy `tf.keras`, you can instead reopen the PR at `keras-team/tf-keras`, which hosts the `TensorFlow-only`, legacy version of `Keras`.
